File: list3/java/Polynomial.java
import java.util.ArrayList;

public class Polynomial<T extends FieldElement<T>> {

  private ArrayList<T> coefficients;
  private int degree;

  private void remove_trailing_zeros() {
    while (this.coefficients.size() > 1 && this.coefficients.get(this.coefficients.size() - 1)._isZero()) {
      this.coefficients.remove(this.coefficients.size() - 1);
    }
    this.degree = this.coefficients.size() - 1;
  }

  public Polynomial(final T[] coefficients) {
    this.coefficients = new ArrayList<T>();
    for (int i = 0; i < coefficients.length; i++) {
      this.coefficients.add(coefficients[i]);
    }
    this.remove_trailing_zeros();
  }

  public Polynomial(final ArrayList<T> coeffs) {
    this.coefficients = coeffs;
    this.remove_trailing_zeros();
  }

  public Polynomial(final Polynomial<T> p) {
    this.coefficients = new ArrayList<T>();
    for (int i = 0; i < p.coefficients.size(); i++) {
      this.coefficients.add(p.coefficients.get(i));
    }

    this.degree = p.degree;
    this.remove_trailing_zeros();
  }

  public Polynomial(final T coeff) {
    this.coefficients = new ArrayList<T>();
    this.coefficients.add(coeff);
    this.degree = 0;
  }

  public int get_degree() {
    return this.degree;
  }

  public T get_coeff(final int i) {
    return this.coefficients.get(i);
  }

  public Polynomial<T> _add(final Polynomial<T> p) {
    ArrayList<T> result = new ArrayList<T>();
    int new_size = this.degree + p.degree + 1;

    for (int i = 0; i < new_size; i++) {
      if (i <= this.degree) {
        if (i <= p.degree) {
          result.add(this.coefficients.get(i)._add(p.coefficients.get(i)));
        } else {
          result.add(this.coefficients.get(i));
        }
      } else if (i <= p.degree) {
        result.add(p.coefficients.get(i));
      }
    }

    return new Polynomial<T>(result);
  }

  public Polynomial<T> sub(final Polynomial<T> p) {
    ArrayList<T> result = new ArrayList<T>();
    int new_size = this.degree + p.degree + 1;

    for (int i = 0; i < new_size; i++) {
      if (i <= this.degree) {
        if (i <= p.degree) {
          result.add(this.coefficients.get(i)._sub(p.coefficients.get(i)));
        } else {
          result.add(this.coefficients.get(i));
        }
      } else if (i <= p.degree) {
        result.add(p.coefficients.get(i)._neg());
      }
    }

    return new Polynomial<T>(result);
  }

  public Polynomial<T> mul(final Polynomial<T> p) {
    ArrayList<T> result = new ArrayList<T>();
    for (int i = 0; i <= this.degree + p.degree; i++) {
      result.add(this.coefficients.get(0)._zero());
    }

    for (int i = 0; i <= this.degree; i++) {
      for (int j = 0; j <= p.degree; j++) {
        result.set(i + j, result.get(i + j)._add(this.coefficients.get(i)._mul(p.coefficients.get(j))));
      }
    }
    return new Polynomial<T>(result);
  }

  public Polynomial<T> mul(final T coeff) {
    ArrayList<T> result = new ArrayList<T>();
    for (int i = 0; i <= this.degree; i++) {
      result.add(this.coefficients.get(i)._mul(coeff));
    }

    return new Polynomial<T>(result);
  }

  public Polynomial<T>[] divmod(final Polynomial<T> p) {
    Polynomial<T>[] result = new Polynomial[2];
    Polynomial<T> quotient = new Polynomial<T>(this.coefficients.get(0)._zero());
    Polynomial<T> remainder = new Polynomial<T>(this);

    while (remainder.degree >= p.degree) {
      T factor = remainder.coefficients.get(remainder.degree)._div(p.coefficients.get(p.degree));
      Polynomial<T> term = new Polynomial<T>(factor);

      for (int i = 0; i < remainder.degree - p.degree; i++) {
        term.coefficients.add(0, term.coefficients.get(0)._zero());
      }
      
      quotient = quotient._add(term);
      remainder = remainder.sub(p.mul(term));
    }

    result[0] = quotient;
    result[1] = remainder;
    return result;
  }

  public Polynomial<T> div(final Polynomial<T> p) {
    return this.divmod(p)[0];
  }

  public Polynomial<T> mod(final Polynomial<T> p) {
    return this.divmod(p)[1];
  }

  public boolean eq(final Polynomial<T> polynomial) {
    if (this.degree != polynomial.degree) {
      return false;
    }

    for (int i = 0; i <= this.degree; i++) {
      if (this.coefficients.get(i)._neq(polynomial.coefficients.get(i))) {
        return false;
      }
    }

    return true;
  }

  public boolean neq(final Polynomial<T> polynomial) {
    return !this.eq(polynomial);
  }

  public boolean lt(final Polynomial<T> polynomial) {
    if (this.degree < polynomial.degree) {
      return true;
    }
    if (this.degree > polynomial.degree) {
      return false;
    }
    for (int i = this.degree; i >= 0; i--) {
      if (this.coefficients.get(i)._lt(polynomial.coefficients.get(i))) {
        return true;
      }

      if (this.coefficients.get(i)._gt(polynomial.coefficients.get(i))) {
        return false;
      }
    }
    return false;
  }

  public boolean gt(final Polynomial<T> polynomial) {
    if (this.degree > polynomial.degree) {
      return true;
    }

    if (this.degree < polynomial.degree) {
      return false;
    }

    for (int i = this.degree; i >= 0; i--) {
      if (this.coefficients.get(i)._gt(polynomial.coefficients.get(i))) {
        return true;
      }

      if (this.coefficients.get(i)._lt(polynomial.coefficients.get(i))) {
        return false;
      }
    }

    return false;
  }

  public boolean le(final Polynomial<T> polynomial) {
    return this.lt(polynomial) || this.eq(polynomial);
  }

  public boolean ge(final Polynomial<T> polynomial) {
    return this.gt(polynomial) || this.eq(polynomial);
  }

  public T value(final T x) {
    T result = this.coefficients.get(0);
    T power = x;
    
    for (int i = 1; i <= this.degree; i++) {
      result = result._add(this.coefficients.get(i)._mul(power));
      power = power._mul(x);
    }
    return result;
  }

  public String toString() {
    String result = "";
    for (int i = 0; i < this.degree; i++) {
      if (this.coefficients.get(i)._isZero())
        continue;

      if (i == 0) {
        result += this.coefficients.get(i).toString();
      } else {
        result += this.coefficients.get(i) + "x^" + i;
      }

      result += " + ";
    }

    result += this.coefficients.get(this.degree).toString();
    return result;
  }

  public boolean equals(Object other) {
    if (other == null || !(other instanceof Polynomial)) {
      return false;
    }

    if (other == this) {
      return true;
    }

    Polynomial<T> other_polynomial = (Polynomial<T>) other;
    if (this.degree != other_polynomial.degree) {
      return false;
    }

    for (int i = 0; i <= this.degree; i++) {
      if (this.coefficients.get(i)._neq(other_polynomial.coefficients.get(i))) {
        return false;
      }
    }

    return true;
  }
}

File: list3/python/polynomial.py
from itertools import zip_longest

def polynomialOver(field: float):
  class Polynomial:
    def __init__(self, coefficients):
      if isinstance(coefficients, Polynomial):
        self.coefficients = coefficients.coefficients
      elif hasattr(coefficients, '__iter__') or hasattr(coefficients, "iter"):
        self.coefficients = [field(x) for x in coefficients]
      else:
        self.coefficients = [field(coefficients)]

      if hasattr(self.coefficients, '__len__') or hasattr(self.coefficients, "len"):
        for i in range(len(self.coefficients) - 1, -1, -1):
          if self.coefficients[i] != field(0):
            break
          self.coefficients.pop(i)      
      
      self.__name__ = f"{field.__name__}[x]"

    def isZero(self):
      return self.coefficients == []

    # String representation
    def __repr__(self):
      if self.isZero():
        return "0"

      poly_repr = ""
      for i, a in enumerate(self.coefficients):
        if i == 0:
          poly_repr += f"{a}"
        elif i == 1:
          poly_repr += f" + {a}x"
        else:
          poly_repr += f" + {a}x^{i}"
      
      return f"{self.__name__} = " + poly_repr

    def __str__(self):
      return repr(self)

    # Polynomial degree methods (len, abs, degree)
    def __len__(self):
      return len(self.coefficients)
    def __abs__(self):
      return len(self.coefficients)
    def degree(self):
      return abs(self) - 1

    # Make Polynomial iterable
    def __iter__(self):
      return iter(self.coefficients)
    def iter(self):
      return self.__iter__()

    # Arithmetic
    @typecast
    def __add__(self, other):
      new_coeffs = [sum(x) for x in zip_longest(self, other, fillvalue=field(0))]
      return Polynomial(new_coeffs)

    def __neg__(self):
      return Polynomial([-a for a in self])

    @typecast
    def __sub__(self, other):
      return self + (-other)

    @typecast
    def __mul__(self, other):
      if self.isZero() or other.isZero():
        return Polynomial(0)
      new_coeffs = [field(0)] * (len(self) + len(other) - 1)
      for i, a in enumerate(self):
        for j, b in enumerate(other):
          new_coeffs[i + j] += a * b
      return Polynomial(new_coeffs)

    @typecast
    def __divmod__(self, div):
      if div.isZero():
        raise ZeroDivisionError("Division by zero")
      if self.degree() < div.degree():
        return Polynomial([]), self
      
      quot, rem = Polynomial([]), self
      div_degree = div.degree()
      div_tail = div.coefficients[-1]

      while rem.degree() >= div_degree:
        degree = rem.degree() - div_degree
        term = [field(0)] * degree
        coeff = [rem.coefficients[-1] / div_tail]
        term = Polynomial(term + coeff)
        quot += term
        rem -= term * div

      return quot, rem

    @typecast
    def __truediv__(self, div):
      quot, _ = divmod(self, div)
      return quot
    
    @typecast
    def __div__(self, div):
      quot, _ = divmod(self, div)
      return quot

    @typecast
    def __mod__(self, div):
      _, rem = divmod(self, div)
      return rem

    # Get polynomial value
    def __call__(self, x):
      return sum([a * (x ** i) for i, a in enumerate(self)])

    # Comparison
    @typecast
    def __eq__(self, other):
      return self.coefficients == other.coefficients

    @typecast
    def __ne__(self, other):
      return self.coefficients != other.coefficients

    @typecast
    def __lt__(self, other):
      if (self.degree() < other.degree()):
        return True
      for i in range(other.degree(), -1, -1):
        if self.coefficients[i] < other.coefficients[i]:
          return True
        elif self.coefficients[i] > other.coefficients[i]:
          return False
      return False

    @typecast
    def __gt__(self, other):
      if (self.degree() > other.degree()):
        return True
      for i in range(other.degree(), -1, -1):
        if self.coefficients[i] > other.coefficients[i]:
          return True
        elif self.coefficients[i] < other.coefficients[i]:
          return False
      return False
  
    @typecast
    def __le__(self, other):
      return self < other or self == other

    @typecast
    def __ge__(self, other):
      return self > other or self == other

    # Get polynomial coefficient
    def __getitem__(self, i):
      if i < 0:
        i += len(self)
      if i < 0 or i >= len(self):
        raise IndexError(f"Index {i} out of range")
      return self.coefficients[i]

    # Set polynomial coefficient
    def __setitem__(self, i, val):
      if i < 0:
        i += len(self)
      if i < 0 or i >= len(self):
        raise IndexError(f"Index {i} out of range")
      self.coefficients[i] = val

    # Right-side arithmetic

    def __radd__(self, other):
      return self + other
    
    def __rsub__(self, other):
      return -self + other

    def __rmul__(self, other):
      return self * other

    def __rdivmod__(self, other):
      return divmod(other, self)

    def __rtruediv__(self, other):
      return other / self

    def __rdiv__(self, other):
      return other / self

  return Polynomial


def typecast(f):
  def __cast(self, other):
    if type(self) is not type(other):
      try:
        other = self.__class__(other) # FieldElement(other)
      except Exception as e:
        raise TypeError(
          f"Couldn't cast {other} of type {type(other).__name__} to type " \
          f"{type(self).__name__} in func={f.__name__}.\n" \
          f"Reason: {e}"
        )
    return f(self, other)
  return __cast


def is_prime(num):
  if num == 2 or num == 3: return True
  if num < 2 or num % 2 == 0: return False
  if num < 9: return True
  if num % 3 == 0: return False
  sqrt_num = int(num ** 0.5) # truncated sqrt
  cand = 5
  while cand <= sqrt_num:
    if num % cand == 0: return False
    if num % (cand+2) == 0: return False
    cand += 6
  return True
